src/sqlfluff/core/parser/segments/base.py

```diff
@@ -1345,156 +1345,169 @@ def apply_fixes(
-
-            # If any fixes applied, do an intermediate reposition. When applying
-            # fixes to children and then trying to reposition them, that recursion
-            # may rely on the parent having already populated positions for any
-            # of the fixes applied there first. This ensures those segments have
-            # working positions to work with.
-            if fixes_applied:
-                seg_buffer = list(
-                    self._position_segments(tuple(seg_buffer), parent_pos=r.pos_marker)
-                )
-
-            # Then recurse (i.e. deal with the children) (Requeueing)
-            seg_queue = seg_buffer
-            seg_buffer = []
-            for seg in seg_queue:
-                s, before, after = seg.apply_fixes(dialect, rule_code, fixes)
-                # 'before' and 'after' will usually be empty. Only used when
-                # lower-level fixes left 'seg' with non-code (usually
-                # whitespace) segments as the first or last children. This is
-                # generally not allowed (see the can_start_end_non_code field),
-                # and these segments need to be "bubbled up" the tree.
-                seg_buffer.extend(before)
-                seg_buffer.append(s)
-                seg_buffer.extend(after)
-
-            # After fixing we should be able to rely on whitespace being
-            # inserted in appropriate places. That logic now lives in
-            # `BaseRule._choose_anchor_segment()`, rather than here.
-
-            # Rather than fix that here, we simply assert that it has been
-            # done. This will raise issues in testing, but shouldn't in use.
-            if (
-                # TODO: Rethink this assertion once parse_grammar is gone.
-                r.parse_grammar
-                and not r.can_start_end_non_code
-                and seg_buffer
-            ):  # pragma: no cover
-                assert not self._find_start_or_end_non_code(seg_buffer), (
-                    "Found inappropriate fix application: inappropriate "
-                    "whitespace positioning. Post `_choose_anchor_segment`. "
-                    "Please report this issue on GitHub with your SQL query. "
-                )
+                            )
+                else:
+                    seg_buffer.append(seg)
+                    seg.set_parent(self)
+            # Invalidate any caches
+            self.invalidate_caches()
+
+        # If any fixes applied, do an intermediate reposition. When applying
+        # fixes to children and then trying to reposition them, that recursion
+        # may rely on the parent having already populated positions for any
+        # of the fixes applied there first. This ensures those segments have
+        # working positions to work with.
+        if fixes_applied:
+            seg_buffer = list(
+                self._position_segments(tuple(seg_buffer), parent_pos=self.pos_marker)
+            )
```

Comment:
This section is unchanged - it's just one less indentation.

Reply:
Definitely review this with the "hide whitespace" github option on, otherwise it's really hard to read.